### ISSUE TITLE
dev-dependencies not installing on python2

### ISSUE THREAD

**Issue by GrimHacker** [enpaul/tox-poetry-installer#68]
I'm trying to use `tox-poetry-installer` to support testing my py2/3 library with `tox` but the dev dependencies are apparently not installing under py2 when run via `tox` (but do when using `poetry`).

I know Python 2.7 has been dead for over 2 years (the requirement to continue support for it at the moment is out of my hands), but since the dependencies install correctly using `poetry` I think its reasonable to expect that they would install correctly using `tox-poetry-installer`

The following can be used to reproduce this issue:

1. Setup a project directory:
```
$ mkdir -p grimlib/grimlib
$ touch grimlib/grimlib.__init__.py
$ touch grimlib/pyproject.toml
$ touch grimlib/tox.ini
```

2. Populate tox.ini and pyproject.ini:

tox.ini:
```
[tox]
isolated_build = true
envlist = {py27,py3}
requires =
    tox-poetry-installer[poetry] == 0.8.3

[testenv]
require_locked_deps = true
install_dev_deps = true

commands =
    python -V
    python -m pylint --version
```

pyproject.toml:
```
[tool.poetry]
name = "grimlib"
version = "1.0.0" 
description = "poetry test"
authors = ["GrimHacker"]

[tool.poetry.dependencies]
python = "^2.7 || ^3.7"

[tool.poetry.dev-dependencies]
pylint = [
    { version = "<=1.9.5", python = "~2.7" },  # dropped py2 after 1.9.5
    { version = ">=2,<3", python = "^3.7" }  # 2.0.0 first version with py3.7
]
pytest = [
    { version = "<5", python = "~2.7" },  # dropped py2 in 5.0
    { version = "<8", python = "^3.7" }
]
lazy-object-proxy = { version = "!=1.7.0"}  # Dep of astroid from pylint. 7.0.0 was yanked but is being selected by poetry

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

3. Lock the dependencies

```
$ poetry lock
Updating dependencies
Resolving dependencies... (18.5s)

Writing lock file
```
4.  Run `tox` and observe the `InvocationError` under py2.
```
$ tox -r
.tox recreate: /home/grimhacker/foo/grimlib/.tox/.tox
.tox installdeps: tox-poetry-installer[poetry] == 0.8.3, tox >= 3.24.5
.package recreate: /home/grimhacker/foo/grimlib/.tox/.package
.package installdeps: poetry-core>=1.0.0
py27 recreate: /home/grimhacker/foo/grimlib/.tox/py27
py27 tox-poetry-installer: Installing 1 dependencies from Poetry lock file (using 10 threads)
py27 inst: /home/grimhacker/foo/grimlib/.tox/.tmp/package/1/grimlib-1.0.0.tar.gz
py27 installed: DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.,grimlib @ file:///home/grimhacker/foo/grimlib/.tox/.tmp/package/1/grimlib-1.0.0.tar.gz,lazy-object-proxy==1.6.0
py27 run-test-pre: PYTHONHASHSEED='2565277340'
py27 run-test: commands[0] | python -V
Python 2.7.17
py27 run-test: commands[1] | python -m pylint --version
/home/grimhacker/foo/grimlib/.tox/py27/bin/python: No module named pylint
ERROR: InvocationError for command /home/grimhacker/foo/grimlib/.tox/py27/bin/python -m pylint --version (exited with code 1)

py3 recreate: /home/grimhacker/foo/grimlib/.tox/py3
py3 tox-poetry-installer: Installing 18 dependencies from Poetry lock file (using 10 threads)
py3 inst: /home/grimhacker/foo/grimlib/.tox/.tmp/package/1/grimlib-1.0.0.tar.gz
py3 installed: astroid==2.9.3,attrs==21.4.0,grimlib @ file:///home/grimhacker/foo/grimlib/.tox/.tmp/package/1/grimlib-1.0.0.tar.gz,iniconfig==1.1.1,isort==5.10.1,lazy-object-proxy==1.6.0,mccabe==0.6.1,packaging==21.3,platformdirs==2.5.1,pluggy==1.0.0,py==1.11.0,pylint==2.12.2,pyparsing==3.0.7,pytest==7.0.1,toml==0.10.2,tomli==2.0.1,typing_extensions==4.1.1,wrapt==1.14.0
py3 run-test-pre: PYTHONHASHSEED='2565277340'
py3 run-test: commands[0] | python -V
Python 3.8.0
py3 run-test: commands[1] | python -m pylint --version
pylint 2.12.2
astroid 2.9.3
Python 3.8.0 (default, Dec  9 2021, 17:53:27) 
[GCC 8.4.0]
______________________________________________________ summary _______________________________________________________
ERROR:   py27: commands failed
  py3: commands succeeded
```

5. Use `poetry` to create a python2.7 virtualenv, install the dependencies, and run `python -m pylint --version` to show the error does not occur:
```
$ poetry env use python2.7
...
$ poetry install
Installing dependencies from lock file

Package operations: 14 installs, 14 updates, 0 removals

  • Updating contextlib2 (0.6.0.post1 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 0.6.0.post1)
  • Updating scandir (1.10.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.10.0)
  • Updating six (1.16.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.16.0)
  • Updating typing (3.10.0.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 3.10.0.0)
  • Updating configparser (4.0.2 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 4.0.2)
  • Updating pathlib2 (2.3.7.post1 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 2.3.7.post1)
  • Updating zipp (1.2.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.2.0)
  • Updating backports.functools-lru-cache (1.6.4 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.6.4)
  • Updating enum34 (1.1.10 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.1.10)
  • Updating futures (3.3.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 3.3.0)
  • Updating importlib-metadata (2.1.3 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 2.1.3)
  • Installing lazy-object-proxy (1.6.0)
  • Updating pyparsing (2.4.7 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 2.4.7)
  • Updating singledispatch (3.7.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 3.7.0)
  • Updating wrapt (1.14.0 /home/grimhacker/.cache/pypoetry/virtualenvs/grimlib-4i1phUPA-py2.7/lib/python2.7/site-packages -> 1.14.0)
  • Installing astroid (1.6.6)
  • Installing atomicwrites (1.4.0)
  • Installing attrs (21.4.0)
  • Installing funcsigs (1.0.2)
  • Installing isort (4.3.21)
  • Installing mccabe (0.6.1)
  • Installing more-itertools (5.0.0)
  • Installing packaging (20.9)
  • Installing pluggy (0.13.1)
  • Installing py (1.11.0)
  • Installing wcwidth (0.2.5)
  • Installing pylint (1.9.5)
  • Installing pytest (4.6.11)

$ poetry run python -m pylint --version
No config file found, using default configuration
__main__.py 1.9.5, 
astroid 1.6.6
Python 2.7.17 (default, Feb 27 2021, 15:10:58) 
[GCC 7.5.0]
```

**Comment by enpaul** [comment]
Hi @GrimHacker thanks for reporting this issue! I agree that supporting installing py2 dependencies should be a goal, even if native support for running on py2 is not (the same way the main poetry project supports it).

I notice you don't add `tox` to the demo project lockfile. Could you share the exact versions of `tox` that you're using? Also, could you check which version of python tox's bootstrapped self-provisioned env is using? Running the below commands from your project directory should show these versions:

```
.tox/.tox/bin/python --version
.tox/.tox/bin/python -m pip freeze
```

**Comment by GrimHacker** [comment]
Hi @enpaul 

```shell
$ .tox/.tox/bin/python --version
Python 3.8.0
```
```shell
$ .tox/.tox/bin/python -m pip freeze
CacheControl==0.12.10
cachy==0.3.0
certifi==2021.10.8
cffi==1.15.0
charset-normalizer==2.0.12
cleo==0.8.1
clikit==0.6.2
crashtest==0.3.1
cryptography==36.0.1
distlib==0.3.4
filelock==3.6.0
html5lib==1.1
idna==3.3
importlib-metadata==4.11.3
jeepney==0.7.1
keyring==23.5.0
lockfile==0.12.2
msgpack==1.0.3
packaging==20.9
pastel==0.2.1
pexpect==4.8.0
pkginfo==1.8.2
platformdirs==2.5.1
pluggy==1.0.0
poetry==1.1.13
poetry-core==1.0.8
ptyprocess==0.7.0
py==1.11.0
pycparser==2.21
pylev==1.4.0
pyparsing==3.0.7
requests==2.27.1
requests-toolbelt==0.9.1
SecretStorage==3.3.1
shellingham==1.4.0
six==1.16.0
toml==0.10.2
tomlkit==0.10.0
tox==3.24.5
tox-poetry-installer==0.8.3
urllib3==1.26.8
virtualenv==20.13.3
webencodings==0.5.1
zipp==3.7.0
```

```shell
$ python3.8 -m tox --version
3.24.5 imported from /home/user/.local/lib/python3.8/site-packages/tox/__init__.py
```

I believe i installed `tox` with `python3.8 -m pip install tox` (on Ubuntu 18.04) and run I run it with `python3.8 -m tox`. I didn't add it to the `pyproject.toml` or `poetry.lock` since I thought that would create a "chicken and egg problem".

I'm new to poetry and using it sensibly in proximity to tox to test under multiple python versions does not seem to be particularly intuitive so I may be doing something wrong... :sweat_smile:

**Comment by enpaul** [comment]
Hi @GrimHacker I'm so sorry for the delay on this (my day job has been very hectic recently) but I finally have a patch up for this problem [here](https://github.com/enpaul/tox-poetry-installer/pull/69). When you get a chance could you take a look at it and let me know if it solves the problem you encountered?

Also as a suggestion for your local environment: I would actually recommend adding both `tox` and `tox-poetry-installer` directly to your Poetry environment, rather than installing them separately or using the Tox `requires` option. Both installing manually to a separate virtualenv and installing using the `requires` option are vulnerable to transient dependency conflicts, which Poetry specifically aims to prevent. Also, if your Tox version isn't managed as part of your lockfile it makes it all the harder to keep the version you're using in sync with that of your peers and/or automation. If you do decide to add them to your Poetry environment then you just need to run `tox` using `poetry run tox`, no chicken-and-egg untangling needed!

**Comment by GrimHacker** [comment]
Unfortunately it doesn't look like I can add `tox-poetry-installer` to `pyproject.toml` since it doesn't support py2.

`pyproject.toml`:
```
[tool.poetry]
name = "grimlib"
version = "1.0.0" 
description = "poetry test"
authors = ["GrimHacker"]

[tool.poetry.dependencies]
python = "^2.7 || ^3.7"

[tool.poetry.dev-dependencies]
tox = "^3.24.5"
tox-poetry-installer = { git = "https://github.com/enpaul/tox-poetry-installer.git", branch = "enp/py2" }
pylint = [
    { version = "<=1.9.5", python = "~2.7" },  # dropped py2 after 1.9.5
    { version = ">=2,<3", python = "^3.7" }  # 2.0.0 first version with py3.7
]
pytest = [
    { version = "<5", python = "~2.7" },  # dropped py2 in 5.0
    { version = "<8", python = "^3.7" }
]
lazy-object-proxy = { version = "!=1.7.0"}  # Dep of astroid from pylint. 7.0.0 was yanked but is being selected by poetry

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

```
$ poetry.lock -vvv
....
  SolverProblemError

  The current project's Python requirement (>=2.7,<3.0 || >=3.7,<4.0) is not compatible with some of the required packages Python requirement:
    - tox-poetry-installer requires Python ^3.6.1, so it will not be satisfied for Python >=2.7,<3.0
  
  Because grimlib depends on tox-poetry-installer (rev enp/py2) which requires Python ^3.6.1, version solving failed.

  at ~/.local/share/pypoetry/venv/lib/python3.6/site-packages/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For tox-poetry-installer, a possible solution would be to set the `python` property to ">=3.7,<4.0"
```

----

I don't seem to be easily able to make tox requires build and install tox-poetry-installer from the git branch either.

**Comment by enpaul** [comment]
Sorry @GrimHacker , I should've clarified: to add `tox-poetry-installer` to your project you'll need to specify that it's only going to be installed when using a python3 development environment.

You can do this in the pyproject like this:
```
tox-poetry-installer = { git = "https://github.com/enpaul/tox-poetry-installer.git", branch = "enp/py2", python = "^3.7" }
```

The docs for this functionality are [here](https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies)

**Comment by enpaul** [comment]
@GrimHacker I've merged #69 and closed this issue for now, but if you update to the new version with the bugfix (`0.8.4`) and are still having problems please feel free to re-open this issue and I'll take another look at it.
